### PR TITLE
Implement unit tests for templates of CassandraDatacenter and Reaper

### PR DESCRIPTION
**Pull request by burmanm**

### CODE REVIEW

**Review by jsanda — CHANGES_REQUESTED** [review]
Did initial review on the first couple Reaper tests. I will add follow up reviews for the Medusa and other tests. 

This is looking awesome :) Thanks!

File: tests/template_cassdc_test.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+			Expect(cassdc.Spec.PodTemplateSpec.Spec.InitContainers[0].Name).To(Equal("jmx-credentials"))
+			// Verify LOCAL_JMX value
+			Expect(len(cassdc.Spec.PodTemplateSpec.Spec.Containers)).To(Equal(1))
+			Expect(len(cassdc.Spec.PodTemplateSpec.Spec.Containers[0].Env)).To(Equal(1))
+			Expect(cassdc.Spec.PodTemplateSpec.Spec.Containers[0].Env[0].Name).To(Equal("LOCAL_JMX"))
+			Expect(cassdc.Spec.PodTemplateSpec.Spec.Containers[0].Env[0].Value).To(Equal("no"))
+		})
+
+		It("disabling reaper", func() {
+			options := &helm.Options{
```
[the remark this answers is not in the document]
Comment (burmanm) [reply tests/template_cassdc_test.go:68]: These are both technically tested in the next section when reaper is disabled again (as only the medusa ones are allowed to be found).

File: tests/template_cassdc_test.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+			renderTemplate(options)
+			Expect(cassdc.Annotations).ShouldNot(HaveKeyWithValue("reaper.cassandra-reaper.io/instance", fmt.Sprintf("%s-reaper-k8ssandra", releaseName)))
+		})
+
+		It("enabling only medusa", func() {
+			options := &helm.Options{
+				SetValues:   map[string]string{"backupRestore.medusa.enabled": "true", "repair.reaper.enabled": "false"},
+				KubectlOptions: defaultKubeCtlOptions,
+			}
```
Comment (jsanda) [inline tests/template_cassdc_test.go:81]: When medusa is enabled we also add volume and volume mounts. We want to verify those as well.

File: tests/template_cassdc_test.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+			options := &helm.Options{
+				SetValues:   map[string]string{"repair.reaper.enabled": "false"},
+				KubectlOptions: defaultKubeCtlOptions,
+			}
+
+			renderTemplate(options)
+			Expect(cassdc.Annotations).ShouldNot(HaveKeyWithValue("reaper.cassandra-reaper.io/instance", fmt.Sprintf("%s-reaper-k8ssandra", releaseName)))
+		})
+
+		It("enabling only medusa", func() {
```
Comment (jsanda) [inline tests/template_cassdc_test.go:77]: When medusa is enabled we also add volume and volume mounts. We want to verify those as well.

File: tests/template_cassdc_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jsanda) [reply tests/template_cassdc_test.go:68]: Good point. I hadn't considered that. One thing though that is missing in the medusa test is a check to verify that `LOCAL_JMX` is not set.

File: tests/template_reaper_test.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+		It("changing datacenter name", func() {
+			targetDcName := "reaper-dc"
+			options := &helm.Options{
+				SetStrValues:   map[string]string{"datacenterName": targetDcName},
+				KubectlOptions: k8s.NewKubectlOptions("", "", "k8ssandra"),
+			}
+
+			renderTemplate(options)
+			// Requires new version of reaper-operator and k8ssandra PR merged
```
Comment (jsanda) [inline tests/template_reaper_test.go:57]: Those changes have been merged.

**Review by jsanda — CHANGES_REQUESTED** [review]

File: tests/template_cassdc_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (burmanm) [reply tests/template_cassdc_test.go:68]: Technically is:

```
Expect(len(cassdc.Spec.PodTemplateSpec.Spec.Containers[0].Env)).To(Equal(1))			Expect(cassdc.Spec.PodTemplateSpec.Spec.Containers[0].Env[0].Name).To(Equal("JVM_EXTRA_OPTS"))
``

There can be only one Env-variable and it has to be JVM_EXTRA_OPTS, so no LOCAL_JMX is possible. But I'll add one more check to the disable reaper that there's no env variables at all.

**Comment by burmanm** [comment]
I could not add uncomment the Autoscheduler part as each time I run `go test -v ./...` it will simply modify my go.mod to this:

``github.com/k8ssandra/reaper-operator v0.0.0-20201204054951-d4834fcfac4d``

No replace, no edit.. nothing helps.

File: tests/template_reaper_test.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+		})
+
+		It("modifying autoscheduling option", func() {
+			options := &helm.Options{
+				SetStrValues:   map[string]string{"autoscheduling": "true"},
+				KubectlOptions: k8s.NewKubectlOptions("", "", "k8ssandra"),
+			}
+
+			renderTemplate(options)
+			//Expect(reaper.Spec.ServerConfig.Autoscheduler).ToNot(BeNil())
```
Comment (jsanda) [inline tests/template_reaper_test.go:67]: I uncommented and changed the property name to `reaper.Spec.ServerConfig.AutoScheduling`.

File: tests/template_reaper_test.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+				KubectlOptions: k8s.NewKubectlOptions("", "", "k8ssandra"),
+			}
+
+			renderTemplate(options)
+			Expect(reaper.Spec.ServerConfig.CassandraBackend.CassandraDatacenter.Name).To(Equal(targetDcName))
+		})
+
+		It("modifying autoscheduling option", func() {
+			options := &helm.Options{
+				SetStrValues:   map[string]string{"autoscheduling": "true"},
```
Comment (jsanda) [inline tests/template_reaper_test.go:62]: The property name here should be `repair.reaper.autoschedule`.

**Review by jsanda — CHANGES_REQUESTED** [review]
I made a couple small changes in `template_reaper_test.go` to get it passing. The reaper-operator dependency in go.mod was not changed.

```
Ran 7 of 7 Specs in 0.644 seconds
SUCCESS! -- 7 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestTests (0.65s)
PASS
ok  	github.com/k8ssandra/k8ssandra/tests	0.683s
```

 I also added a Makefile. I pushed my two commits to my [fork](https://github.com/jsanda/k8ssandra/tree/template_unit_tests). 

Can you incorporate these changes?

**Review by jeffbanks — APPROVED** [review]
Overall LGTM.

One comment, there may be some collaboration and sharing of the [`test util module`](https://github.com/k8ssandra/k8ssandra/blob/f90a29de0d19eab0838663a223e54ab6769b240e/tests/integration/util/lib.go) functions applicable to unit test `setup` (i.e. setup and management of options).  

We could construct a similar module for `unit tests` w/ our standardized set of Terratest, Ginkgo, and Omega APIs.

**Review by jsanda — APPROVED** [review]

**Pull request merged**